### PR TITLE
Allow arbitrary python code

### PR DESCRIPTION
**Pull request by Maleware**
# Description

This PR is the outcome of the decision https://github.com/stackabletech/decisions/issues/21
 
Contains:
- CONFIG_OVERRIDE_FILE_HEADER_KEY (operator-rs constant)
- CONFIG_OVERRIDE_FILE_FOOTER_KEY (operator-rs constant)
-  Bump stackable-operator from 0.73.0 to 0.74.0
- Removing `..Default::default()` from lib.rs as we don't want default values ( can be forgotten )
- Adding Error for `writeln!` 

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxe

```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [ ] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] Code contains useful logging statements
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs/style-guide).
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

File: rust/operator-binary/src/superset_controller.rs [hunk 1]
```diff
@@ -532,6 +548,13 @@ fn build_rolegroup_config_map(
         rolegroup: rolegroup.clone(),
     })?;
 
+    // By removing the key from `config_properties`, we avoid pasting the Python code into a Python variable as well
+    // (which would be bad)
+    // TODO: Use public constants in operator-rs to reference `EXPERIMENTAL_FILE_HEADER` and `EXPERIMENTAL_FILE_FOOTER`
```
Comment (sbernauer) [inline rust/operator-binary/src/superset_controller.rs:553]:
```suggestion
```

File: rust/operator-binary/src/superset_controller.rs [hunk 2]
```diff
@@ -523,6 +529,16 @@ fn build_rolegroup_config_map(
     );
 
     let mut config_file = Vec::new();
+
+    // By removing the key from `config_properties`, we avoid pasting the Python code into a Python variable as well
+    // (which would be bad)
+    // TODO: Use public constants in operator-rs to reference `EXPERIMENTAL_FILE_HEADER` and `EXPERIMENTAL_FILE_FOOTER`
+    if let Some(header) = config_properties.remove("EXPERIMENTAL_FILE_HEADER") {
```
Comment (sbernauer) [inline rust/operator-binary/src/superset_controller.rs:536]: Please put this into a constant, ideally in operator-rs
```
// See https://github.com/stackabletech/decisions/issues/21
pub const CONFIG_OVERRIDE_FILE_HEADER_KEY: &str = "EXPERIMENTAL_FILE_HEADER";

// same for footer
```
E.g. here https://github.com/stackabletech/operator-rs/blob/37e91d6440625e174228880eccf287b9d8665654/crates/stackable-operator/src/product_config_utils.rs#L10

**Review by sbernauer — CHANGES_REQUESTED** [review]

File: rust/operator-binary/src/superset_controller.rs (hunk 1 above)
Comment (Maleware) [reply rust/operator-binary/src/superset_controller.rs]: I've removed a part of the comments.

I would leave the `TODO:` until we can use the consts from `operator-rs`

File: rust/operator-binary/src/superset_controller.rs (hunk 2 above)
Comment (Maleware) [reply rust/operator-binary/src/superset_controller.rs:534]: Will be done with https://github.com/stackabletech/operator-rs/pull/843

File: rust/operator-binary/src/superset_controller.rs (hunk 1 above)
Comment (sbernauer) [reply rust/operator-binary/src/superset_controller.rs]: I think the comment is valuable and would prefer to keep it ;)
I just meant the duplicated comment below, where we actually don't `remove` anything ;)

Can you please re-add the first comment again?

Comment (Maleware) [reply rust/operator-binary/src/superset_controller.rs]: Damn it 😂

First thing tomorrow morning!

File: rust/operator-binary/src/superset_controller.rs (hunk 2 above)
Comment (Maleware) [reply rust/operator-binary/src/superset_controller.rs]: Done with https://github.com/stackabletech/superset-operator/pull/530/commits/21c328bdb0b60bf7642c0fc9f3c735eff24b9ce6

File: rust/operator-binary/src/superset_controller.rs (hunk 1 above)
Comment (Maleware) [reply rust/operator-binary/src/superset_controller.rs]: Done with https://github.com/stackabletech/superset-operator/pull/530/commits/057039d9830443204dc422325d4a6d20e3387fc3

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,13 +2,19 @@
...
+### Added
+
+- Allowing arbitrary python code as EXPERIMENTAL_FILE_HEADER and EXPERIMENTAL_FILE_FOOTER in superset_config.py ([#530])
+
 ### Changed
 
 - Reduce CRD size from `472KB` to `45KB` by accepting arbitrary YAML input instead of the underlying schema for the following fields ([#528]):
   - `podOverrides`
   - `affinity`
+- Bumping stackable-operator from 0.73.0 to 0.74.0 ([#530])
```
Comment (sbernauer) [inline CHANGELOG.md:14]: Sorry, we don't track that :see_no_evil: 
```suggestion
```

File: rust/operator-binary/src/superset_controller.rs [hunk 4]
```diff
@@ -523,6 +532,13 @@ fn build_rolegroup_config_map(
     );
 
     let mut config_file = Vec::new();
+
```
Comment (sbernauer) [inline rust/operator-binary/src/superset_controller.rs:535]:
```suggestion

    // By removing the keys from `config_properties`, we avoid pasting the Python code into a Python variable as well
    // (which would be bad)
```

File: rust/operator-binary/src/superset_controller.rs [hunk 5]
```diff
@@ -523,6 +532,13 @@ fn build_rolegroup_config_map(
...
+    if let Some(header) = config_properties.remove(CONFIG_OVERRIDE_FILE_HEADER_KEY) {
+        writeln!(config_file, "{}", header).context(WriteToConfigFileStringSnafu)?;
+    }
+    // removing key from `config_properties` to avoid key value match. Append it later.
```
Comment (sbernauer) [inline rust/operator-binary/src/superset_controller.rs:539]:
```suggestion
```

**Review by sbernauer — CHANGES_REQUESTED** [review]
Sorry ^^
I still think some docs would be great (with the usual experimental disclaimer)

**Comment by Maleware** [comment]
Excellent point, done with https://github.com/stackabletech/superset-operator/pull/530/commits/4b64920519cc82444e04536023f232d7cacade5f

**Review by razvan — COMMENTED** [review]
Some tests would be good.  Especially testing if role group and role configOverrides work as expected.

**Review by sbernauer — DISMISSED** [review]
LGTM % Razvans comment

**Comment by Maleware** [comment]
🟢 `./scripts/run-tests --test smoke_superset-4.0.2_openshift-false --skip-release `

```
--- PASS: kuttl (89.84s)
    --- PASS: kuttl/harness (0.00s)
        --- PASS: kuttl/harness/smoke_superset-4.0.2_openshift-false (89.83s)
PASS

```

File: tests/templates/kuttl/smoke/30-install-superset.yaml.j2 [hunk 6]
```diff
@@ -36,6 +36,14 @@ spec:
     config:
       logging:
         enableVectorAgent: {{ lookup('env', 'VECTOR_AGGREGATOR') | length > 0 }}
+    configOverrides:
+      superset_config.py:
```
Comment (razvan) [inline tests/templates/kuttl/smoke/30-install-superset.yaml.j2:40]: Hm, looks like there was a misunderstanding. I meant testing explicitely that `EXPERIMENTAL_FILE_HEADER` and his friend can be overridden with `configOverrides` even if they are special cased by the operator.

The general case is already tested in #531 and it's no need to be repeated here.

File: tests/templates/kuttl/smoke/32-assert.yaml [hunk 7]
```diff
@@ -0,0 +1,12 @@
+---
```
Comment (razvan) [inline tests/templates/kuttl/smoke/32-assert.yaml:1]: I would rename this to 31-assert.yaml. Better to resolve a conflict than testing the same thing twice.

**Review by razvan — APPROVED** [review]
approved as discussed

**Comment by lfrancke** [comment]
The decision is not public, can you summarize what this is about and if it is something that is documented please include a link to the docs. If it's not documented can you briefly summarize why note?

Does this need to be mentioned in the release notes?

**Comment by Maleware** [comment]
This is to allow the user to add arbitrary python code to the `superset_config.py`. 

We've got it documented in our [superset docs](https://docs.stackable.tech/home/nightly/superset/usage-guide/configuration-environment-overrides#_configuration_properties) and [airflow docs](https://docs.stackable.tech/home/nightly/airflow/usage-guide/overrides#_configuration_properties)

For a customer we added a whole class and reference to it to this config file s.th. they were able to use a custom `OIDC` login/logut.
However, this can be used for any case where superset allows us to replace functions with custom ones. It was driven by [superset documentation about custom OIDC manager](https://superset.apache.org/docs/configuration/configuring-superset/#keycloak-specific-configuration-using-flask-oidc).

We've reached a decision on that this is how superset wants to be configured ( no matter how ugly it might appear ) and thus we want the operator to be useful in this case.
 
We've had an additional customer for airflow who wrote their whole config for themself ( basically ignoring the operators work and thus it's not used ) to achieve having own classes and functions. They now can use the configs the operator provides and have their custom code as a config. 

I can't really make a objective statement weather or not we should mention it in the release notes. For me it should as it enables some customers and they might wanna know.

Left to do: We have a [draft PR](https://github.com/stackabletech/docker-images/pull/799) in docker images where we *can* include necessary libraries to use OIDC with superset properly ( airflow might wanna have that too, to be discussed ).

**Comment by lfrancke** [comment]
Thank you. In that case: Can you please add a short sentence as it could appear in the release notes as a comment here?

**Comment by Maleware** [comment]
Sure

Release Notes:
Apache Superset: Allow custom arbitrary python code in superset_config.py
